src/System/IO/BlockIO.hs

```diff
@@ -184,52 +187,112 @@ data IOOp m = IOOpRead  !Fd !FileOffset !(MutableByteArray (PrimState m)) !Int !
 --
 submitIO :: IOCtx -> V.Vector (IOOp IO) -> IO (VU.Vector IOResult)
 submitIO IOCtx {
+           ioctxBatchSizeLimit',
            ioctxQSemN,
            ioctxURing,
            ioctxChanIOBatch,
            ioctxChanIOBatchIx
          }
-         ioops = do
-    let !iobatchOpCount = V.length ioops
-    waitQSemN ioctxQSemN iobatchOpCount
-    iobatchIx         <- readChan ioctxChanIOBatchIx
-    iobatchCompletion <- newEmptyMVar
-    let iobatchKeepAlives = ioops
-    writeChan ioctxChanIOBatch
-              IOBatch {
-                iobatchIx,
-                iobatchOpCount,
-                iobatchCompletion,
-                iobatchKeepAlives
-              }
-    withMVar ioctxURing $ \case
-      Nothing -> throwIO closed
-      Just uring -> do
---      print ("submitIO", iobatchOpCount)
-        V.iforM_ ioops $ \ioopix ioop ->
-            let !ioopid = packIOOpId iobatchIx ioopix
-            in
-              --print ioop >>
-              case ioop of
-              IOOpRead  fd off buf bufOff cnt -> do
-                guardPinned buf
-                URing.prepareRead  uring fd off
-                                  (mutableByteArrayContents buf `plusPtr` bufOff)
-                                  cnt ioopid
-              IOOpWrite fd off buf bufOff cnt -> do
-                guardPinned buf
-                URing.prepareWrite uring fd off
-                                  (mutableByteArrayContents buf `plusPtr` bufOff)
-                                  cnt ioopid
-        URing.submitIO uring
---      print ("submitIO", "submitting done")
-    takeMVar iobatchCompletion
+         ioops
+    | iobatchOpCount == 0
+    = return VU.empty
+
+    | iobatchOpCount > ioctxBatchSizeLimit'
```

Comment:
This looks reasonable, but it also looks like we could share almost all of this code between the single and multiple batch case. And this is all sufficiently long and complicated that avoiding code duplication here is probably the way to go.